tests/lang/s02_core/s01_procedures/t02_argument_passing.nim

```diff
@@ -0,0 +1,318 @@
+discard """
+description: '''
+Different ways of passing arguments to procedures.
+'''
+"""
+
+##
+
+
+block different_number_or_arguments:
+  ## Procedure can accept zero or more arguments.
+  proc zero(): int = 1
+
+  doAssert zero() == 1
+
+  ## Arguments are specified in the procedure declaration
+  proc single(arg: int): int = arg
+
+  doAssert single(12) == 12
+  doAssert single(2) == 2
+
+  ## Multiple arguments can be specified either separately
+  proc multi2(arg1: int, arg2: int): (int, int) = (arg1, arg2)
+
+  doAssert multi2(1, 2) == (1, 2)
+  doAssert multi2(3, 4) == (3, 4)
+
+  ## Or together, if they have the same type
+  proc multi3(arg1, arg2: int): (int, int) = (arg1, arg2)
+
+  doAssert multi3(1, 2) == (1, 2)
+  doAssert multi3(3, 4) == (3, 4)
+
+# Putting different calling syntaxes in the middle of different argument
+# modes because it is necessary to have one or more arguments passed, but
+# is not really important
+block regular_syntax:
+  ## Procedure can be called with different
+  proc regular(arg: int) = discard
+
+  regular(12)
+
+
+block command_syntax:
+  proc command(arg: int) = discard
+
+  command 12
+
+  ## Note that due to UFCS whitespace separating function name from the
+  ## arguments is important. In this case we have two procedures - one that
+  ## accepts tuple, and other for two arguments.
+  proc takesTuple(arg: (int, int)) = discard
+  proc takesTwoArguments(arg1, arg2: int) = discard
+
+  ## Adding whitespace before parens will be considered calling proc via
+  ## command syntax and passing tuple as an argument.
+  takesTuple (1, 2)
+
+  ## No leading whitespace in this case will be considered a call with
+  ## regular syntax
+  takesTwoArguments(1, 2)
+
+block block_syntax:
+  ## If needed, expression for arguments can be placed on the
+  ## multiple lines using block call syntax.
+  proc blockProc(arg: int) = discard
+
+  ## Block call syntax using parens
+  blockProc():
+    12
+
+  ## Block call with command syntax
+  blockProc:
+    12
+
+  proc blockReturn(arg: int): int = arg
+
+  ## Block syntax for procedure calls can also be used as an expression,
+  ## although no in the infix expressions (like `doAseert 12 == <block here>`)
+  block:
+    let res = blockReturn():
+      12
+
+    doAssert res == 12
+
+  block:
+    let res = blockReturn:
+      12
+
+    doAssert res == 12
+
+block block_syntax_do:
+  ## If there are multiple arguments that have to be placed on the
+  ## separate lines this can be achieved using `do:` for
+  ## the arguments.
+  proc blockProc(arg1: int, arg2: int): (int, int) = (arg1, arg2)
```

Comment:
Ain't no party like a blockProcParty, because a blockProcParty can do all the arguments.